### DISCUSSION TITLE
Get method signature location (line/column numbers) through mono.cecil

### DISCUSSION THREAD

**Discussion by xi-liu-ds** [jbevain/cecil#857] [General]
Hi,

As it is easy to get the location of method body instructions through `MethodDefinition.DebugInformation.SequencePoints` of `mono.cecil`, I am wondering can we get the location, such as line number and column number, of method name and signature through `mono.cecil`?

Thanks!

**Comment by huynhit24** [comment]
You can check it out this way.
[Source](https://stackoverflow.com/questions/54469446/how-to-get-line-number-of-instruction-using-mono-cecil-0-10-2)

**Comment by xi-liu-ds** [comment]
> You can check it out this way. [Source](https://stackoverflow.com/questions/54469446/how-to-get-line-number-of-instruction-using-mono-cecil-0-10-2)

Thanks for your prompt response! However, I was actually asking if we can get the location of **method signature or method name**, like what is indicated in the attached code snippet below.

```cs
using System;

namespace HelloWorld
{
  class Program
  {
    static void Main(string[] args)        // <- location of this part of the code
    {
      Console.WriteLine("Hello World!");    
    }
  }
}
```

**Comment by Zastai** [comment]
Unfortunately, no. The MethodDef table in the CLR metadata doesn't store source information, and the Portable PDB metadata's method debug info is based on IL offsets within methods ([docs](https://github.com/dotnet/runtime/blob/main/docs/design/specs/PortablePdb-Metadata.md)). The sequence points, if present, are stored in offset order, so if there are sequence points, the first one will be the one _closest_ to the method declaration, but that's the best you can do.

**Comment by Zastai** [comment]
_Usually_ (i.e. if it's the result of a normal C# compilation using the Microsoft compiler), certainly in a Debug build (not 100% about Release builds), there'll be a sequence point attached to a `nop` instruction at the top for the `{` of the method.

So if you use the standard MS style with the `{` on a line of its own, then you could have a basic heuristic of:
- if there are sequence points
- the method starts with a `nop` which has a sequence point
- that sequence point describes a single character
- then the method declaration will be one line before that (obviously, longer signatures may start earlier)

Looking now, constructors seem different; they don't get a `nop`, but they start with an invocation of `Object::.ctor` which has a sequence point covering the signature.

Of course, none of the above is _guaranteed_ in any way.

**Comment by ltrzesniewski** [comment]
Methods start with a `nop` only in debug builds for instance 🙂
